ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -595,33 +595,36 @@ private void readFromFileAndPersistIfNecessary(String mountPath) throws Replicat
               FindToken token = factory.getFindToken(stream);
               // update token
               PartitionInfo partitionInfo = partitionsToReplicate.get(partitionId);
-              boolean updatedToken = false;
-              for (RemoteReplicaInfo remoteReplicaInfo : partitionInfo.getRemoteReplicaInfos()) {
-                if (remoteReplicaInfo.getReplicaId().getDataNodeId().getHostname().equalsIgnoreCase(hostname)
-                    && remoteReplicaInfo.getReplicaId().getDataNodeId().getPort() == port
-                    && remoteReplicaInfo.getReplicaId().getReplicaPath().equals(replicaPath)) {
-                  logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname,
-                      port, token);
-                  if (partitionInfo.getStore().getSizeInBytes() > 0) {
-                    remoteReplicaInfo.initializeTokens(token);
-                    remoteReplicaInfo.setTotalBytesReadFromLocalStore(totalBytesReadFromLocalStore);
-                  } else {
-                    // if the local replica is empty, it could have been newly created. In this case, the offset in
-                    // every peer replica which the local replica lags from should be set to 0, so that the local
-                    // replica starts fetching from the beginning of the peer. The totalBytes the peer read from the
-                    // local replica should also be set to 0. During initialization these values are already set to 0,
-                    // so we let them be.
-                    tokenWasReset = true;
-                    replicationMetrics.replicationTokenResetCount.inc();
-                    logger.info("Resetting token for partition {} remote host {} port {}, persisted token {}",
-                        partitionId, hostname, port, token);
+              if (partitionInfo != null) {
```

Comment:
It seems like it is probably safest to always reset token/persist whenever a store is found to be offline. However, I wonder if there are any situations where a store being unavailable would not require the token to be reset? If there are, we would incur the cost of checking blobs from the start of each of the remote replicas.